File: src/lib/tiposkripto/src/BaseAction.ts
import type { TestTypeParams_any } from "./CoreTypes.js";

/**
 * BaseAction is the internal unified base class for all action phases.
 * It covers BDD's When, AAA's Act, and TDT's Feed.
 * This class is not exposed to users - use BaseWhen, BaseShould, or BaseIt instead.
 */
export abstract class BaseAction<I extends TestTypeParams_any> {
  public name: string;
  actionCB: (x: I["iselection"]) => I["then"];
  error: Error = null as any;
  artifacts: string[] = [];
  status: boolean | undefined;

  addArtifact(path: string) {
    if (typeof path !== "string") {
      throw new Error(
        `[ARTIFACT ERROR] Expected string, got ${typeof path}: ${JSON.stringify(
          path,
        )}`,
      );
    }
    const normalizedPath = path.replace(/\\/g, "/");
    this.artifacts.push(normalizedPath);
  }

  constructor(name: string, actionCB: (xyz: I["iselection"]) => I["then"]) {
    this.name = name;
    this.actionCB = actionCB;
  }

  abstract performAction(
    store: I["istore"],
    actionCB: (x: I["iselection"]) => I["then"],
    testResource: any,
    artifactory?: any,
  ): Promise<any>;

  toObj() {
    const obj = {
      name: this.name,
      status: this.status,
      error: this.error
        ? `${this.error.name}: ${this.error.message}\n${this.error.stack}`
        : null,
      artifacts: this.artifacts,
    };
    return obj;
  }

  async test(store: I["istore"], testResourceConfiguration: any, artifactory?: any) {
    try {
      const result = await this.performAction(
        store,
        this.actionCB,
        testResourceConfiguration,
        artifactory,
      );
      this.status = true;
      return result;
    } catch (e: any) {
      this.status = false;
      this.error = e;
      throw e;
    }
  }
}

export type IActions<I extends TestTypeParams_any> = Record<
  string,
  BaseAction<I>
>;

File: src/lib/kafe/src/main/java/kafe/BaseAction.java
package kafe;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.HashMap;
import java.util.function.Function;

/**
 * BaseAction is the unified base class for all action phases.
 * It covers BDD's When, AAA's Act, and TDT's Feed.
 */
public abstract class BaseAction<S, R> {
    public String name;
    public Function<R, R> actionCB;
    public Exception error;
    public List<String> artifacts;
    public Boolean status;
    
    public BaseAction(String name, Function<R, R> actionCB) {
        this.name = name;
        this.actionCB = actionCB;
        this.artifacts = new ArrayList<>();
        this.status = null;
    }
    
    public void addArtifact(String path) {
        if (!(path instanceof String)) {
            throw new RuntimeException(
                "[ARTIFACT ERROR] Expected string, got " + path.getClass().getName() + ": " + path
            );
        }
        String normalizedPath = path.replace('\\', '/');
        artifacts.add(normalizedPath);
    }
    
    public abstract R performAction(
        R store,
        Function<R, R> actionCB,
        ITTestResourceConfiguration testResourceConfiguration,
        Object artifactory
    );
    
    public Map<String, Object> toObj() {
        String errorStr = null;
        if (error != null) {
            errorStr = error.getClass().getName() + ": " + error.getMessage();
        }
        Map<String, Object> obj = new HashMap<>();
        obj.put("name", name);
        obj.put("status", status);
        obj.put("error", errorStr);
        obj.put("artifacts", artifacts);
        return obj;
    }
    
    public R test(R store, ITTestResourceConfiguration testResourceConfiguration) throws Exception {
        return test(store, testResourceConfiguration, null);
    }
    
    public R test(R store, ITTestResourceConfiguration testResourceConfiguration, Object artifactory) throws Exception {
        try {
            R result = performAction(store, actionCB, testResourceConfiguration, artifactory);
            status = true;
            return result;
        } catch (Exception e) {
            status = false;
            error = e;
            throw e;
        }
    }
}
